atomix/cluster/src/test/java/io/atomix/raft/ControllableRaftContexts.java

```diff
@@ -0,0 +1,351 @@
+import org.jmock.lib.concurrent.DeterministicScheduler;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Uses a DeterministicScheduler and controllable messaging layer to get a deterministic execution
+ * of raft threads. Note:- Currently there is some non-determinism hidden in the raft. Hence the
+ * resulting execution is not fully deterministic.
+ */
+public class ControllableRaftContexts {
+
+  private final Map<MemberId, ControllableRaftServerProtocol> serverProtocols = new HashMap<>();
+  private final Map<MemberId, Queue<Tuple<Runnable, CompletableFuture>>> messageQueue =
+      new HashMap<>();
+  private final Map<MemberId, DeterministicSingleThreadContext> deterministicExecutors =
+      new HashMap<>();
+
+  private Path directory;
+
+  private final int nodeCount;
+  private final Map<MemberId, RaftContext> raftServers = new HashMap<>();
+  private Duration electionTimeout;
+  private Duration hearbeatTimeout;
+  private int nextEntry = 0;
+
+  // Used only for verification. Map[term -> leader]
+  private final Map<Long, MemberId> leadersAtTerms = new HashMap<>();
+
+  public ControllableRaftContexts(final int nodeCount) {
+    this.nodeCount = nodeCount;
+  }
+
+  public Map<MemberId, RaftContext> getRaftServers() {
+    return raftServers;
+  }
+
+  public RaftContext getRaftContext(final int memberId) {
+    return raftServers.get(MemberId.from(String.valueOf(memberId)));
+  }
+
+  public RaftContext getRaftContext(final MemberId memberId) {
+    return raftServers.get(memberId);
+  }
+
+  public void setup(final Path directory, final Random random) throws Exception {
+    this.directory = directory;
+    if (nodeCount > 0) {
+      createRaftContexts(nodeCount, random);
+    }
+    joinRaftServers();
+    electionTimeout = getRaftContext(0).getElectionTimeout();
+    hearbeatTimeout = getRaftContext(0).getHeartbeatInterval();
+
+    // expecting 0 to be the leader
+    tickHeartbeatTimeout(0);
+  }
+
+  public void shudown() throws IOException {
+    raftServers.forEach((m, c) -> c.close());
+    raftServers.clear();
+    serverProtocols.clear();
+    deterministicExecutors.forEach((m, e) -> e.close());
+    deterministicExecutors.clear();
+    messageQueue.clear();
+    leadersAtTerms.clear();
+    directory = null;
+  }
+
+  private void joinRaftServers() throws InterruptedException, ExecutionException, TimeoutException {
+    final Set<CompletableFuture<Void>> futures = new HashSet<>();
+    final var servers = getRaftServers();
+    final var serverIds = new ArrayList<>(servers.keySet());
+    final long electionTimeout =
+        servers.get(MemberId.from(String.valueOf(0))).getElectionTimeout().toMillis();
+    Collections.sort(serverIds);
+    servers.forEach(
+        (memberId, raftContext) -> futures.add(raftContext.getCluster().bootstrap(serverIds)));
+
+    runUntilDone(0);
+    // trigger election on 0 so that 0 is initially the leader
+    getDeterministicScheduler(MemberId.from(String.valueOf(0)))
```

Comment:
Couldn't we introduce in our normal code that node one for partition one does immediately a poll? Last time we had the direct transition, which was not really safe.

Reply:
We can do. But that could also lead to a leader election whenever a node (re)starts. So when a leader dies, a new leader is elected, when the old leader comes back (after a few seconds/minutes) there will be an election again which may or may not result in a new leader.

Reply:
But for that we have the poll mechanism right? So first it starts a poll and this should be rejected if the other followers have received recently heart beats right?

Reply:
If this regarding balancing leader across replicas, can we discuss it separately?